05.wc/wc.rb

```diff
@@ -1,67 +1,57 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'debug'
 
-opt = OptionParser.new
-option = {}
-opt.on('-l') { |l| option[:l] = l }
-opt.on('-w') { |w| option[:w] = w }
-opt.on('-c') { |c| option[:c] = c }
-argv = opt.parse(ARGV)
+def main
+  _opt, option, argv = set_option
 
-def option_or_no_option(file, option, argv)
-  option_l = file.lines.count.to_s.rjust(8)
-  option_w = file.split.size.to_s.rjust(8)
-  option_c = file.size.to_s.rjust(8)
+  if argv[0].nil?
+    deta = $stdin.read
+    option_l, option_w, option_c, show_count = a(deta, argv, option)
 
-  print option_l if option[:l]
-  print option_w if option[:w]
-  print option_c if option[:c]
+  else
+    option_l_total_num = 0
+    option_w_total_num = 0
+    option_c_total_num = 0
 
-  if option == {}
-    print option_l
-    print option_w
-    print option_c
-  end
-
-  print " #{argv[0]}\n"
-end
+    argv.each do |file_name|
+      deta = File.read(file_name)
+      option_l, option_w, option_c, show_count = a(deta, file_name, option)
 
-if argv[0].nil?
-  file = []
-  loop do
-    file_name = $stdin.gets
-    break if file_name.nil?
+      option_l_total_num += option_l
+      option_w_total_num += option_w
+      option_c_total_num += option_c
+    end
 
-    file.push(file_name.to_s)
+    print_count(option_l_total_num, option_w_total_num, option_c_total_num, option)
+    puts ' total'
   end
-  file = file.join
-
-  option_or_no_option(file, option, argv)
-
-else
-  option_l_total_num = 0
-  option_w_total_num = 0
-  option_c_total_num = 0
-
-  argv.each do |file_name|
-    file = File.read(file_name)
-
-    option_or_no_option(file, option, argv)
+end
 
-    option_l_total_num += file.lines.count
-    option_w_total_num += file.split.size
-    option_c_total_num += file.size
-  end
+def print_count(option_l, option_w, option_c, option)
+  print option_l.to_s.rjust(8) if option[:l] || option == {}
```

Comment:
```suggestion
  print option_l.to_s.rjust(8) if option[:l] || option.empty?
```

とした方が効率が良さそうです。
`{}`を使うと毎回ハッシュのインスタンスが作られるので。